golem-component-service-base/src/service/component.rs

```diff
@@ -304,6 +413,142 @@ impl ComponentServiceDefault {
+        <Namespace as TryFrom<String>>::Error: Display + Debug + Send + Sync + 'static,
+    {
+        let versioned_component_id = create_new_versioned_component_id(component_id);
+
+        // analyze component before uploading anything so we fail early
+        let component_metadata = ComponentMetadata::analyse_component(&data)
+            .map_err(ComponentError::ComponentProcessingError)?;
+
+        let component_size = data.len() as u64;
+
+        info!(namespace = %namespace,"Uploaded component - exports {:?}", component_metadata.exports);
+
+        let (_, _) = tokio::try_join!(
+            self.upload_user_component(&versioned_component_id, data.clone()),
+            self.upload_protected_component(&versioned_component_id, data)
+        )?;
+
+        let component = Component {
+            component_size,
+            metadata: component_metadata,
+            created_at: Utc::now(),
+            component_type,
+            component_name: component_name.clone(),
+            versioned_component_id: versioned_component_id.clone(),
+            namespace: namespace.clone(),
+            files: uploaded_files.clone(),
+        };
+
+        let record = component
+            .clone()
+            .try_into()
+            .map_err(|e| ComponentError::conversion_error("record", e))?;
+
+        let result = self.component_repo.create(&record).await;
+
+        if let Err(RepoError::UniqueViolation(_)) = result {
+            Err(ComponentError::AlreadyExists(component_id.clone()))?;
+        }
+
+        self.component_compilation
+            .enqueue_compilation(component_id, component.versioned_component_id.version)
+            .await;
+
+        Ok(component)
+    }
+
+    async fn update_unchecked<Namespace>(
+        &self,
+        component_id: &ComponentId,
+        data: Vec<u8>,
+        component_type: Option<ComponentType>,
+        files: Option<Vec<InitialComponentFile>>,
+        namespace: &Namespace,
+    ) -> Result<Component<Namespace>, ComponentError>
+    where
+        Namespace: Display + TryFrom<String> + Eq + Clone + Send + Sync,
+        <Namespace as TryFrom<String>>::Error: Display + Debug + Send + Sync + 'static,
+    {
+        let created_at = Utc::now();
+
+        let metadata = ComponentMetadata::analyse_component(&data)
+            .map_err(ComponentError::ComponentProcessingError)?;
+
+        info!(namespace = %namespace, "Uploaded component - exports {:?}", metadata.exports);
+
+        let constraints = self.component_repo.get_constraint(component_id).await?;
+
+        let new_type_registry = FunctionTypeRegistry::from_export_metadata(&metadata.exports);
+
+        if let Some(constraints) = constraints {
+            let conflicts =
+                Self::find_component_metadata_conflicts(&constraints, &new_type_registry);
+            if !conflicts.is_empty() {
+                return Err(ComponentError::ComponentConstraintConflictError(conflicts));
+            }
+        }
```

Comment:
Is this repeated anywhere else? May be in `update` method?

Reply:
There are there things now for both update and create:
* update: This is used by the rest api and uploads all files to the blob_storage + checks consistency of the archive and permissions list. Afterwards it's forwarded to update_unchecked
* update_internal: This is used by the grpc api and just checks that all referenced files are in the blob_store, then it forwards to update_unchecked
* update_uncheck: Does the actual logic of uploading files. Everything that is duplicated between the above two code paths lives in here.

I admit the update_internal name is a bit bad, suggestions welcome :)

Reply:
I am thinking. @vigoo @nicoburniske @noise64  May be jump in for some suggestions if you have any. 
I hope this won't block the merge though. After the explanation I understand why you did this.

Reply:
I think the names are "fine". 
I see the logic in having these two different ways of updating in the REST vs gRPC APIs, and right now it does not really matter because the component update gRPC API is not really used (only by the test framework). 

We did not really do such separate behaviors in REST vs gRPC (I think we not necessarily treat the gRPC API as internal, we are just not exposing it by default in our setup). That said it makes sense to have more internal endpoints on the gRPC APIs, like this one, and once we have it, we can always add anoter gRPC endpoint which incorporates the upload too (like the REST one), if needed.

So for me this is not a blocking issue and I'm good with adding it like this.